site/docs/types/type_system.md

```diff
@@ -0,0 +1,36 @@
+
+To illustrate, Substrait defines an `i8` to be an integer between -128 and 127, but an implementation may use any physical storage format that can represent (at least) that range of integers. The most likely physical representation for that would be a single byte represented as two's complement, but any of these would be valid, too (if increasingly ill-advised):
+
+ - a 32-bit two's-complement integer;
+ - a 9-bit sign-magnitude number;
+ - a single-precision float (since all integers in that range can be exactly represented);
+ - the decimal string representation of the number;
+
+and so on. The behavior of a plan involving physical values out of the logical range of a type is unspecified, and therefore consumer-dependent. Furthermore, if a physical storage format has more than one representation for a single logical value, such as (1 year, 0 months) and (0 years, 12 months) for `interval_year`, the particular representation returned after a Substrait consumer operates on value is unspecified and should not be relied upon.
+
+For each logical type, Substrait also defines a simple, canonical physical representation. This representation is used for literal expressions, virtual tables, and so on. Note that this representation is not intended to be particularly performant or storage-efficient.
+
+Logical types are subdivided into three classes:
+
+ - [simple types](simple_logical_types.md): builtin logical types without parameters;
+ - [compound types](compound_logical_types.md): builtin logical types that need to be parameterized; and
+ - [user-defined types](user_defined_types.md): types defined by means of [simple extensions](../extensions/index.md#simple-extensions).
+
+## Null Handling
+
+All types in Substrait must have a declared nullability, either `REQUIRED` to specify non-nullable or `NULLABLE` to specify nullable. In the latter case, values of the type may assume the special null value, otherwise they may not. When represented as text, a type is `NULLABLE` if and only if the logical type name is suffixed with a `?`, for example `i8?` or `VARCHAR?<10>` (see [type syntax parsing](type_parsing.md)).
+
+[Unless otherwise specified](../expressions/scalar_functions.md#nullability-handling), operations that take multiple arguments support any mix of nullability at their input. They yield a nullable type if any of the arguments are nullable, and at runtime yield null if any argument is null.
+
+Null literals in Substrait are typed: for example, a null for an `i8?` is different from a null for a `VARCHAR?<10>`. There is no separate null type.
+
+## Coercion and Casting
+
+The only context in which Substrait may be argued to perform type coercion is for nullability, as specified in the previous section. In all other contexts, explicit [cast expressions](../expressions/casts.md) are needed.
```

Comment:
This isn't a property of type, it is a property of the function definition system. As such, it is a little weird it is noted here.

Reply:
I disagree strongly here. Yes, it's not a property of *a* type, but coercion (or less generally promotion) is one of the most important properties of a type *system*. It's one of the first questions I ask myself when I encounter a new one.

Even in Substrait, functions merely exist in the context of a type system. Some counterexamples to this purely being a function definition thing:

 - If-else expressions. One of my FIXMEs in the validator is due to nullability behavior there not being specified. I was strict about requiring equal nullability for the branches and non-nullable conditions initially, but IIRC I got tired of manually correcting Isthmus-generated plans when making the test cases.
 - Filter conditions. If the condition is nullable, do all columns in the schema become nullable? Is it illegal to do that? Does it throw an error at runtime if a null comes by? Or does null just imply false? Again, I ended up being lenient here in the validator.
 - Equality functions, which are currently assumed to exist as part of joins, aggregations, set stuff, etc. I don't remember off the top of my head what I did here in the validator.

Whenever this behavior is unspecified, a producer can't accurately deduce the return type of things without making assumptions. Overload resolution depends on nullability, so it needs to be able to do this deduction for the same reasons a read relation has to specify a schema; as soon as a type isn't 100% concrete, it can't know exactly which functions can operate on it. Exhaustively specifying this behavior for everything that takes an expression what the type must be, all the way down to nullability, will only lead to confusion and inconsistency. These two sentences would solve all these problems at once with what seems to me to be reasonably sane default behavior.

I'd also be fine with stating that coercion or promotion doesn't exist, and that types always need to match exactly (unless handled via function nullability handling, I guess). Personally I'd even prefer that. But my gut says it would break a lot of existing stuff.

Reply:
> coercion (or less generally promotion) is one of the most important properties of a type system

If your statement is that "a non-nullable type instance can be promoted to a nullable type instance without loss of information", I agree with that statement. However, the specific behavior of things like you list need to specifically defined. Wrt to your specific statements:

> If-else expressions. One of my FIXMEs in the validator is due to nullability behavior there not being specified. I was strict about requiring equal nullability for the branches and non-nullable conditions initially, but IIRC I got tired of manually correcting Isthmus-generated plans when making the test cases.

Agree that we need to better specify. This was my intention: 
* All branches must return the same type of data ignoring nullability. 
* All branches must return the same type variations.
* If all of the branches are required, the outcome is required. If any one of the branches is nullable, the output is nullable.

> Filter conditions. If the condition is nullable, do all columns in the schema become nullable? Is it illegal to do that? Does it throw an error at runtime if a null comes by? Or does null just imply false? Again, I ended up being lenient here in the validator.

I don't understand this question. A filter relation has no impact on the schema of the data. If the condition returns null, we should behave the same as databases do which I believe is to exclude the given record (we can't confirm that the filter condition passes).

> Equality functions, which are currently assumed to exist as part of joins, aggregations, set stuff, etc. I don't remember off the top of my head what I did here in the validator.

Agree this needs to be better defined.

Reply:
> If your statement is that "a non-nullable type instance can be promoted to a nullable type instance without loss of information", I agree with that statement. However, the specific behavior of things like you list need to specifically defined.

I'm using the word "promotion" to mean a something that happens automatically, i.e. a coercion, that is guaranteed to not lose information. For example, if I have a function that only takes nullable arguments, and i have a non-nullable type, I can still pass it to that function without an explicit cast, due to this coercion rule.

From this discussion I think your intentions however are that coercions should never happen. Which is fine, even preferable IMO. However, by the current specification if-else branches must then have exactly equal types, including nullability:

> When an if expression is declared, all return expressions must be the same identical type.

and since the spec says nothing about the expected type for the various conditions, a strict validator should probably reject anything that isn't a non-nullable boolean with the default variation. The "implicit equality functions" should then also return that type rather than being viral about nulls, i.e. null == 1 equals false instead of null. Again, that's perfectly consistent, I was just worried it might be *too* strict.

> I don't understand this question. A filter relation has no impact on the schema of the data. If the condition returns null, we should behave the same as databases do which I believe is to exclude the given record (we can't confirm that the filter condition passes).

If that's what databases do, then fine. I honestly wouldn't know, so I don't really have an opinion about what the behavior should be, other than that it needs to be specified. By the coercion logic I proposed, one might argue that a nullable filter condition makes the columns in the schema nullable and yields a row with all nulls if the condition yields null, which seems to be in line with the viral nature of null I see everywhere else. When the default logic is that conditions must be non-nullable unless otherwise specified there would also be no confusion, it'd just need an extra note mentioning that null means false in the context of a condition to get the behavior you describe.

Reply:
>  However, by the current specification if-else branches must then have exactly equal types, including nullability:

I think we need to allow coercion specifically in those locations. 

> I don't really have an opinion about what the behavior should be

It some ways, this is great. It's a huge test of the specification. It's very rigorous. This is also a highly inefficient way of trying to formalize the specification. It'd be great if more people helped who had a common ground of the SQL specification and the great amount of commonalities the vast majority of these systems carry (and Substrait is just trying to reflect).